codec/src/types/map.rs

```diff
@@ -0,0 +1,270 @@
+//! Codec implementations for HashMap.
+//!
+//! For portability and consistency between architectures,
+//! the size of the map must fit within a [`u32`].
+
+use crate::{
+    codec::{EncodeSize, Read, Write},
+    error::Error,
+    varint, Config, RangeConfig,
+};
+use bytes::{Buf, BufMut};
+use std::{collections::HashMap, hash::Hash};
+
+// Write implementation for HashMap
+impl<K: Ord + Hash + Eq + Write, V: Write> Write for HashMap<K, V> {
+    fn write(&self, buf: &mut impl BufMut) {
+        let len = u32::try_from(self.len()).expect("HashMap length exceeds u32::MAX");
+        varint::write(len, buf);
+
+        // Sort the keys to ensure deterministic encoding
+        let mut keys: Vec<_> = self.keys().collect();
+        keys.sort();
+        for key in keys {
+            key.write(buf);
+            self.get(key).unwrap().write(buf);
+        }
+    }
+}
+
+// EncodeSize implementation for HashMap
+impl<K: Ord + Hash + Eq + EncodeSize, V: EncodeSize> EncodeSize for HashMap<K, V> {
+    fn encode_size(&self) -> usize {
+        // Start with the varint size of the length
+        let len = u32::try_from(self.len()).expect("HashMap length exceeds u32::MAX");
+        let mut size = varint::size(len);
+
+        // Add the encoded size of each key and value
+        // Note: Iteration order doesn't matter for size calculation.
+        for (key, value) in self {
+            size += key.encode_size();
+            size += value.encode_size();
+        }
+        size
+    }
+}
+
+// Read implementation for HashMap
+impl<
+        R: RangeConfig,
+        KCfg: Config,
+        VCfg: Config,
+        K: Read<KCfg> + Clone + Ord + Hash + Eq,
+        V: Read<VCfg> + Clone,
+    > Read<(R, (KCfg, VCfg))> for HashMap<K, V>
+{
+    fn read_cfg(
+        buf: &mut impl Buf,
+        (range, (k_cfg, v_cfg)): &(R, (KCfg, VCfg)),
+    ) -> Result<Self, Error> {
+        // Read and validate the length prefix
+        let len32 = varint::read::<u32>(buf)?;
+        let len = usize::try_from(len32).map_err(|_| Error::InvalidVarint)?;
+        if !range.contains(&len) {
+            return Err(Error::InvalidLength(len));
+        }
+        let mut map = HashMap::with_capacity(len);
+
+        // Keep track of the last key read
+        let mut last_key: Option<K> = None;
+
+        // Read each key-value pair
+        for _ in 0..len {
+            let key = K::read_cfg(buf, k_cfg)?;
+
+            // Check if keys are in ascending order relative to the previous key
+            if let Some(ref last) = last_key {
+                if key <= *last {
```

Comment:
Do we want to return a different error for duplicate keys vs non-ascending keys?

Reply:
If not, I suggest modifying the error to make it clearer that it isn't quite ascending